### PR TITLE
Improve how AI chooses damaging moves

### LINKED ISSUE #3130: Possible issue with AI move choice on killing rolls
### Description

Currently the AI picks a random move between all moves that see a kill, including moves with lower accuracy. (Somewhat discussed in PR #3109)
pokeemerald-8

### Version

upcoming (Edge)

### Upcoming Version

_No response_

### Discord contact info

_No response_

### LINKED ISSUE #3153: AI isn't picking the best move possible to faint the target
### Description

If both moves are super effective the AI picks one randomly without considering the damage the moves are doing. 

pokeemerald-20
pokeemerald-19

### Version

upcoming (Edge)

### Upcoming Version

_No response_

### Discord contact info

_No response_

### PR DESCRIPTION
**Pull request by DizzyEggg**
Fixes #3130
Fixes #3153

The general flow should be like this(assuming there are no special effects which would otherwise be more preferable):
if can faint -> choose the move without side effects(e.g. recoil, recharge, loading turn), and with better accuracy(e.g. Tackle over Mega Punch if both are 1hko)
if can't faint -> choose the move which takes the least no of hits to KO, but only if the other move doesn't have any advantages(otherwise it's a toss-up and depends on the battle conditions).

Also fixed Taunt's score to be minus 10 and improved handling of Solar Beam.

Before:
pokeemerald_01

254561947-0587a7ef-6651-4172-ba75-3bda57df9eeb

After:
pokeemerald_06

pokeemerald_08

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
What do you think about removing this switch block?
```
        switch (AI_DATA->effectiveness[battlerAtk][battlerDef][AI_THINKING_STRUCT->movesetIndex])
        {
        case AI_EFFECTIVENESS_x8:
            score += 8;
            break;
        case AI_EFFECTIVENESS_x4:
            score += 4;
            break;
        case AI_EFFECTIVENESS_x2:
            if (AI_RandLessThan(176))
                score += 2;
            else
                score++;
            break;
        }
```
Your PR takes care of a lot of edge cases and this block looks seems unnecessary with it. You would just get score increases for the sake of increasing it.

**Comment by AlexOn1ine** [comment]
I did some testing and my suggestion generally would be to only decrease the score when compare to a better move. 
The problem with increases is that it inflates the score which might render status move unusable.
An example with the current behavior of this PR
```
Good move, score 101
Bad Move, score 101 
Bad Move, score 101
Bad Move, score 101 
```
Let's say all moves have a score of 101 before their comparison.

**Comment by AlexOn1ine** [comment]
I did some testing and my suggestion generally would be to only decrease the score when compare to a better move.
The problem with increases is that it inflates the score which might render status move unusable.
Let's say we have the base 
```
Good move, score 101
Bad Move, score 101 
Bad Move, score 101
Bad Move, score 101
```
If the comparison from this PR are applied it ends up like this
```
Good move, score 104
Bad Move, score 100 
Bad Move, score 100
Bad Move, score 100
```
My suggestion would be to decrease the bad moves only
```
Good move, score 101
Bad Move, score 100 
Bad Move, score 100
Bad Move, score 100
```
Same result but you don't inflate the score

**Comment by DizzyEggg** [comment]
Despite the issue you Alex mentioned I still think this PR does more good than evil. Would it be fine to merge it to upcoming and then improve it? The thing is even if it's not perfect with this PR, it is _still_ worse without it. Also, I hope to have AI tests soon, so it'll be easier to catch these issues.

**Comment by DizzyEggg** [comment]
@AsparagusEduardo Should be ready to merge.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -3122,6 +3137,18 @@ static bool32 IsPinchBerryItemEffect(u32 holdEffect)
...
+static u32 CompareMoveAccuracies(u32 battlerAtk, u32 battlerDef, u32 moveSlot1, u32 moveSlot2)
+{
+    u32 acc1 = AI_DATA->moveAccuracy[battlerAtk][battlerDef][moveSlot1];
+    u32 acc2 = AI_DATA->moveAccuracy[battlerAtk][battlerDef][moveSlot2];
+
+    if (acc1 > acc2)
+        return 0;
+    else if (acc2 > acc1)
+        return 1;
+    return 2;
```
[the remark this answers is not in the document]
Comment (DizzyEggg) [reply src/battle_ai_main.c:3149]: Which constants?

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**